locator/tablets.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "locator/tablet_replication_strategy.hh"
+#include "locator/tablets.hh"
+#include "types/types.hh"
+#include "types/tuple.hh"
+#include "types/set.hh"
+#include "utils/hash.hh"
+#include "db/system_keyspace.hh"
+#include "cql3/query_processor.hh"
+#include "cql3/untyped_result_set.hh"
+#include "replica/database.hh"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+namespace locator {
+
+seastar::logger tablet_logger("tablets");
+
+std::ostream& operator<<(std::ostream& out, const tablet_replica& r) {
+    return out << r.host << ":" << r.shard;
+}
+
+const tablet_map& tablet_metadata::get_tablet_map(table_id id) const {
+    try {
+        return _tablets.at(id);
+    } catch (const std::out_of_range&) {
+        throw std::runtime_error(format("Tablet map not found for table {}", id));
+    }
+}
+
+tablet_map& tablet_metadata::get_tablet_map(table_id id) {
+    return const_cast<tablet_map&>(
+        const_cast<const tablet_metadata*>(this)->get_tablet_map(id));
+}
+
+void tablet_metadata::set_tablet_map(table_id id, tablet_map map) {
+    _tablets[id] = std::move(map);
+}
+
+future<> tablet_metadata::clear_gently() {
+    for (auto&& [id, map] : _tablets) {
+        co_await map.clear_gently();
+    }
+    co_return;
+}
+
+tablet_map::tablet_map()
+        : tablet_map(1)
+{ }
+
+tablet_map::tablet_map(size_t tablet_count)
+        : _log2_tablets(log2ceil(tablet_count)) {
+    tablet_count = 1ul << _log2_tablets;
+    _tablets.resize(tablet_count);
+}
+
+const tablet_info& tablet_map::get_tablet_info(tablet_id id) const {
+    return _tablets[id];
+}
+
+tablet_id tablet_map::get_tablet_id(token t) const {
+    if (t.is_minimum()) {
+        return 0;
+    }
+    if (t.is_maximum()) {
+        return tablet_count() - 1;
+    }
```

Comment:
doesn't `dht::compaction_group_of` take care of the above cases?
if not, then it should be fixed.
I don't see why `get_tablet_id` and `compaction_group_of` should diverge.